imblearn/over_sampling/_smote.py

```diff
@@ -1334,3 +1336,247 @@ def _fit_resample(self, X, y):
+
+    ADASYN : Over-sample using ADASYN.
+
+    KMeansSMOTE : Over-sample applying a clustering before to oversample using
+        SMOTE.
+
+    References
+    ----------
+    .. [1]  Ramentol, E., Caballero, Y., Bello, R. et al. SMOTE-RSB *: a
+       hybrid preprocessing approach based on oversampling and undersampling
+       for high imbalanced data-sets using SMOTE and rough sets theory. Knowl
+       Inf Syst 33, 245–265 (2012). https://doi.org/10.1007/s10115-011-0465-6.
+
+    Examples
+    --------
+    >>> from collections import Counter
+    >>> from sklearn.datasets import make_classification
+    >>> from imblearn.over_sampling import SMOTERSB
+    >>> X, y = make_classification(n_classes=2, class_sep=2,
+    ... weights=[0.1, 0.9], n_informative=3, n_redundant=1, flip_y=0,
+    ... n_features=20, n_clusters_per_class=1, n_samples=1000, random_state=10)
+    >>> print('Original dataset shape %s' % Counter(y))
+    Original dataset shape Counter({{1: 900, 0: 100}})
+    >>> sm = SMOTERSB(random_state=42)
+    >>> X_res, y_res = sm.fit_resample(X, y)
+    >>> print('Resampled dataset shape %s' % Counter(y_res))
+    Resampled dataset shape Counter({{0: 900, 1: 900}})
+    """
+
+    @_deprecate_positional_args
+    def __init__(
+        self,
+        *,
+        sampling_strategy="auto",
+        random_state=None,
+        k_neighbors=2,
+        n_jobs=None,
+        similarity_threshold=0.4,
+        threshold_increment=0.05,
+        equivalence_set=None,
+        similarity_func=None,
+    ):
+        super().__init__(
+            sampling_strategy=sampling_strategy,
+            random_state=random_state,
+            k_neighbors=k_neighbors,
+            n_jobs=n_jobs,
+        )
+        self.similarity_threshold = similarity_threshold
+        self.threshold_increment = threshold_increment
+        self.equivalence_set = equivalence_set
+        self.similarity_func = similarity_func
+
+    def _validate_params(self, X):
+        if not (0.1 <= self.similarity_threshold <= 0.9):
+            raise ValueError(
+                "`similarity threshold` must lie in the interval [0, 0.9)"
+            )
+        if self.equivalence_set is None:
+            self.feature_indices_ = range(X.shape[1])
+        elif not isinstance(self.equivalence_set, list):
+            raise TypeError("`equivalence_set` must be a list of indices")
+        elif np.any(max(self.equivalence_set) - 1 >= X.shape[1] - 1):
+            raise ValueError("`An index of `equivalence_set` is out of bounds")
+        else:
+            self.feature_indices_ = self.equivalence_set
+
+        if (self.similarity_func is not None and
+                not isinstance(self.similarity_func, Callable)):
+            raise TypeError("`similarity_func` must be a callable")
+
+    # VERY slow! cython might be better suited for this function
+    def _make_similarity_matrix(self, X_s, X_m, maxmin_diff):
```

Comment:
I look quickly at the paper. It seems that the similarity is used to find some neighbours.
I am not entirely sure but a `NearestNeighbors` with a given radius (i.e. the similarity-value (distance normalized by the max distance) would be a way to get a faster implementation. Managing categorical and numerical value would be to implement different distance (e.g. preocmputed `ValueDifferenceMetric` as in `SMOTEN`.